lib/src/Pos/Block/Base.hs

```diff
@@ -0,0 +1,169 @@
+       , HasHeaderHash BlockHeader
+       , HasConfiguration
+       )
+
+-- | Smart constructor for 'MainBlockHeader'.
+mkMainHeader
+    :: SanityConstraintM
+    => Maybe BlockHeader
+    -> SlotId
+    -> SecretKey
+    -> ProxySKBlockInfo
+    -> Body MainBlockchain
+    -> MainExtraHeaderData
+    -> MainBlockHeader
+mkMainHeader prevHeader slotId sk pske body extra =
+    -- here we know that header creation can't fail, because the only invariant
+    -- which we check in 'verifyBBlockHeader' is signature correctness, which
+    -- is enforced in this function
+    leftToPanic "mkMainHeader: " $
+    mkGenericHeader prevHeader body consensus extra
+  where
+    difficulty = maybe 0 (succ . view difficultyL) prevHeader
+    makeSignature toSign (Left psk) =
+        BlockPSignatureLight $ proxySign SignMainBlockLight sk psk toSign
+    makeSignature toSign (Right (psk,_)) =
+        BlockPSignatureHeavy $ proxySign SignMainBlockHeavy sk psk toSign
+    signature prevHash proof =
+        let toSign = MainToSign prevHash proof slotId difficulty extra
+        in maybe
+               (BlockSignature $ sign SignMainBlock sk toSign)
+               (makeSignature toSign)
+               pske
+    leaderPk = maybe (toPublic sk) (either pskIssuerPk snd) pske
+    consensus prevHash proof =
+        MainConsensusData
+        { _mcdSlot = slotId
+        , _mcdLeaderKey = leaderPk
+        , _mcdDifficulty = difficulty
+        , _mcdSignature = signature prevHash proof
+        }
+
+-- | Smart constructor for 'MainBlock'. Uses 'mkMainHeader'. It
+-- verifies consistency of given data and may fail.
+mkMainBlock
+    :: (HasUpdateConfiguration, SanityConstraintM, MonadError Text m)
+    => Maybe BlockHeader
+    -> SlotId
+    -> SecretKey
+    -> ProxySKBlockInfo
+    -> Body MainBlockchain
+    -> m MainBlock
+mkMainBlock prevHeader slotId sk pske body =
+    recreateGenericBlock
+        (mkMainHeader prevHeader slotId sk pske body extraH)
+        body
+        extraB
+  where
+    extraB :: MainExtraBodyData
+    extraB = MainExtraBodyData (mkAttributes ())
+    extraH :: MainExtraHeaderData
+    extraH =
+        MainExtraHeaderData
+            lastKnownBlockVersion
+            curSoftwareVersion
+            (mkAttributes ())
+            (hash extraB)
+
+-- | Empty (i. e. no payload) body of main block for given local slot index.
+emptyMainBody ::
+       HasConfiguration
+    => LocalSlotIndex
+    -> Body MainBlockchain
+emptyMainBody slot =
+    MainBody
+    { _mbTxPayload = emptyTxPayload
+    , _mbSscPayload = defaultSscPayload slot
+    , _mbDlgPayload = def
+    , _mbUpdatePayload = def
+    }
+
+----------------------------------------------------------------------------
+-- Genesis smart constructors
+----------------------------------------------------------------------------
+
+type SanityConstraintG
+     = ( HasDifficulty BlockHeader
+       , HasHeaderHash BlockHeader
+       , HasConfiguration
+       )
```

Comment:
This one is the same as `SanityConstraintM`, I don't think it's a good idea to separate them.